ompi/proc/proc.h

```diff
@@ -369,16 +369,42 @@ static inline bool ompi_proc_is_sentinel (ompi_proc_t *proc)
     return (intptr_t) proc & 0x1;
 }
 
-static inline intptr_t ompi_proc_name_to_sentinel (opal_process_name_t name)
+/*
+ * we assume an ompi_proc_t is at least aligned on two bytes,
+ * so if the LSB of a pointer to an ompi_proc_t is 1, we have to handle
+ * this pointer as a sentinel instead of a pointer.
+ * a sentinel can be seen as a 64 bits array with the following format :
+ * - bit  0     : 1
+ * - bits 1-15  : local jobid
+ * - bits 16-31 : job family
+ * - bits 32-63 : vpid
```

Comment:
Let me try to be more concrete, and offer an a less fragile alternative. The issue here is that we are tying this entire logic to the rather arbitrary fact that an opal_process_name_t currently has the same size as a 64-bit pointer. This is definitely _not_ guaranteed to be true in the future. Past proposals have extended that size, and there are even some discussions about decreasing it to 32-bits. So making the assumption that an opal_process_name_t is 64-bits in size is a bad one.

The problem is essentially caused by us putting all the ompi_proc_t's in a single array, regardless of what job they come from. So you can't use the index of the array as representing the vpid of a specific process in a given job. This is easily rectified in a more flexible manner by simply changing the ompi_proc_t array into an array of arrays.

In other words, create an array of ompi_proc_t's for each jobid with which we are connected, and have the "header" for the array include the jobid. The overwhelming majority of applications never cross-connect between jobs or call comm_spawn, and so we will only rarely see more than one such array anyway. Now each location in the array can be uniquely correlated to a vpid as the vpid is just the index of that location.

You no longer require a sentinel in this situation as you can just use NULL to indicate that we haven't yet defined the ompi_proc_t for that vpid. If the location is NULL, then you construct the process name using the jobid in the array's header and the index of that location.

This more cleanly solves the problem, doesn't break 32-bit environments, and future-proofs us. It doesn't hurt us in any way as the communicators all just carry pointers to their participating ompi_proc_t's - there is no requirement that those objects occupy the same array.

Make sense?